mangaki/mangaki/admin.py

```diff
@@ -229,60 +230,103 @@ def make_nsfw(self, request, queryset):
 
     make_nsfw.short_description = "Rendre NSFW les œuvres sélectionnées"
 
-    # FIXME : https://github.com/mangaki/mangaki/issues/205
     def update_tags_via_anidb(self, request, queryset):
-        if request.POST.get("post"):
-            kept_ids = request.POST.getlist('checks')
-            for anime_id in kept_ids:
-                anime = Work.objects.get(id=anime_id)
-
-                retrieved_tags = anime.retrieve_tags(client)
-                deleted_tags = retrieved_tags["deleted_tags"]
-                added_tags = retrieved_tags["added_tags"]
-                updated_tags = retrieved_tags["updated_tags"]
-
-                tags = deleted_tags
-                tags.update(added_tags)
-                tags.update(updated_tags)
-
-                client.update_tags(anime, tags)
+        works = queryset.all()
 
-            self.message_user(request, "Modifications sur les tags faites")
+        if request.POST.get('confirm'): # Updating tags has been confirmed
+            to_update_work_ids = set(list(map(int, request.POST.getlist('to_update_work_ids'))))
+            nb_updates = len(to_update_work_ids)
+
+            work_ids = list(map(int, request.POST.getlist('work_ids')))
+            tag_titles = request.POST.getlist('tag_titles')
+            tag_weights = list(map(int, request.POST.getlist('weights')))
+            tag_anidb_tag_ids = list(map(int, request.POST.getlist('anidb_tag_ids')))
+            tag_operations = request.POST.getlist('tag_operations')
+
+            # These checkboxes make it possible to know which tags should be added, kept, modified or removed
+            tag_checkboxes = request.POST.getlist('tag_checkboxes')
+            tags_to_process = [list(map(int, tag_checkbox.split(':'))) for tag_checkbox in tag_checkboxes]
+
+            # Build a dict that links a work ID and the dict of tags
+            final_tags_with_work_id = {}
+            for index, work_id in enumerate(work_ids):
```

Comment:
Is there any reason to not iterate over `tags_to_process` which would save a `O(N)` lookup where `N = len(tags_to_process)` for each `work_id`?
If it's because you need the `index` to access structured data, I'd suggest to (in any order):
1) recombine the different tag_stuff into an tags with a Tag (namedtuple or whatever), make dicts where you need.
2) use a `dict` for `tags_to_process` (use as key, something like `{{work_id}}_{{anidb_tag_id}}`).
3) it's actually okay to do two loops to process "to delete works" and "to add/modify/keep works" if that simplify your logic / improve the performance.